In [2]:
import pandas as pd

In [3]:
df=pd.read_csv("C:\projects\Healthcare-Analytics-Dashboard\python\data\medications.csv")

In [4]:
df.head(5)

,START,STOP,PATIENT,PAYER,ENCOUNTER,CODE,DESCRIPTION,BASE_COST,PAYER_COVERAGE,DISPENSES,TOTALCOST,REASONCODE,REASONDESCRIPTION
0,2019-10-30,2019-11-13,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,7c4411ce-02f1-39b5-b9ec-dfbea9ad3c1a,8bca6d8a-ab80-4cbf-8abb-46654235f227,308182,Amoxicillin 250 MG Oral Capsule,7.05,0.0,1,7.05,NaN,NaN
1,2019-10-30,2019-11-13,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,7c4411ce-02f1-39b5-b9ec-dfbea9ad3c1a,8bca6d8a-ab80-4cbf-8abb-46654235f227,313820,Acetaminophen 160 MG Chewable Tablet,5.72,0.0,1,5.72,NaN,NaN
2,2020-02-12,2020-02-26,067318a4-db8f-447f-8b6e-f2f61e9baaa5,5059a55e-5d6e-34d1-b6cb-d83d16e57bcf,adedca64-700b-4fb9-82f1-9cbb658abb73,313820,Acetaminophen 160 MG Chewable Tablet,5.25,0.0,1,5.25,NaN,NaN
3,2020-04-28,2020-05-08,067318a4-db8f-447f-8b6e-f2f61e9baaa5,5059a55e-5d6e-34d1-b6cb-d83d16e57bcf,e03b96de-5604-4989-a2d5-03a63e041eab,834061,Penicillin V Potassium 250 MG Oral Tablet,13.90,0.0,1,13.90,43878008.0,Streptococcal sore throat (disorder)
4,2010-11-22,2011-08-30,ae9efba3-ddc4-43f9-a781-f72019388548,d47b3510-2895-3b70-9897-342d681c769d,11a2dfae-53d4-4d13-a74b-c540a525a1c4,999967,amLODIPine 5 MG / Hydrochlorothiazide 12.5 MG ...,263.49,0.0,9,2371.41,59621000.0,Hypertension


In [5]:
df.shape

(431262, 13)

In [6]:
df.columns

Index(['START', 'STOP', 'PATIENT', 'PAYER', 'ENCOUNTER', 'CODE', 'DESCRIPTION',
       'BASE_COST', 'PAYER_COVERAGE', 'DISPENSES', 'TOTALCOST', 'REASONCODE',
       'REASONDESCRIPTION'],
      dtype='str')

In [7]:
df.dtypes

START                    str
STOP                     str
PATIENT                  str
PAYER                    str
ENCOUNTER                str
CODE                   int64
DESCRIPTION              str
BASE_COST            float64
PAYER_COVERAGE       float64
DISPENSES              int64
TOTALCOST            float64
REASONCODE           float64
REASONDESCRIPTION        str
dtype: object

In [8]:
df.isna().sum()

START                     0
STOP                  28291
PATIENT                   0
PAYER                     0
ENCOUNTER                 0
CODE                      0
DESCRIPTION               0
BASE_COST                 0
PAYER_COVERAGE            0
DISPENSES                 0
TOTALCOST                 0
REASONCODE           100130
REASONDESCRIPTION    100130
dtype: int64

In [9]:
df['START'] = pd.to_datetime(df['START'], errors='coerce')
df['STOP'] = pd.to_datetime(df['STOP'], errors='coerce')

In [10]:
df.head(1)

,START,STOP,PATIENT,PAYER,ENCOUNTER,CODE,DESCRIPTION,BASE_COST,PAYER_COVERAGE,DISPENSES,TOTALCOST,REASONCODE,REASONDESCRIPTION
0,2019-10-30,2019-11-13,f0f3bc8d-ef38-49ce-a2bd-dfdda982b271,7c4411ce-02f1-39b5-b9ec-dfbea9ad3c1a,8bca6d8a-ab80-4cbf-8abb-46654235f227,308182,Amoxicillin 250 MG Oral Capsule,7.05,0.0,1,7.05,NaN,NaN


In [11]:
df[['START', 'STOP']].dtypes

START    datetime64[us]
STOP     datetime64[us]
dtype: object

In [12]:
df.duplicated().sum()

np.int64(3)

In [13]:
df.duplicated(subset=['PATIENT', 'START', 'DESCRIPTION', 'ENCOUNTER']).sum()  #This checks for repeated medication records for the same patient, start date, medication, and encounter.

np.int64(4017)

In [14]:
df=df.drop_duplicates()

In [15]:
df.duplicated().sum()

np.int64(0)

In [16]:
df[['BASE_COST', 'PAYER_COVERAGE', 'DISPENSES', 'TOTALCOST']].describe()

,BASE_COST,PAYER_COVERAGE,DISPENSES,TOTALCOST
count,431259.000000,431259.000000,431259.000000,4.312590e+05
mean,262.227083,93.871241,11.840555,2.626298e+03
std,452.942249,362.667409,44.199548,1.733621e+04
min,0.460000,0.000000,1.000000,4.600000e-01
25%,82.610000,0.000000,1.000000,2.634900e+02
50%,263.490000,0.000000,9.000000,7.329900e+02
75%,263.490000,163.490000,12.000000,3.161880e+03
max,7017.570000,6977.570000,1144.000000,1.886359e+06


In [17]:
df[['BASE_COST', 'PAYER_COVERAGE', 'DISPENSES', 'TOTALCOST']].min()

BASE_COST         0.46
PAYER_COVERAGE    0.00
DISPENSES         1.00
TOTALCOST         0.46
dtype: float64

In [18]:
df[['BASE_COST', 'PAYER_COVERAGE', 'DISPENSES', 'TOTALCOST']].max()

BASE_COST            7017.57
PAYER_COVERAGE       6977.57
DISPENSES            1144.00
TOTALCOST         1886359.20
dtype: float64

Investigate the extreme cost records(10 most expensive medication records.)

In [19]:
df.nlargest(10,"TOTALCOST")[['PATIENT', 'DESCRIPTION', 'START', 'STOP',
     'DISPENSES', 'BASE_COST', 'PAYER_COVERAGE', 'TOTALCOST']]

,PATIENT,DESCRIPTION,START,STOP,DISPENSES,BASE_COST,PAYER_COVERAGE,TOTALCOST
20527,0100f99a-1b5d-4a5b-a73f-559a920412e5,10 ML oxaliplatin 5 MG/ML Injection,1965-12-26,NaT,660,2858.12,0.0,1886359.20
76447,9d43a914-56cf-463f-8535-a0c0c95fc4a8,10 ML oxaliplatin 5 MG/ML Injection,1971-07-06,NaT,595,3110.91,0.0,1850991.45
193768,c78f5852-5ce3-408b-a355-cf04c0ca635b,10 ML oxaliplatin 5 MG/ML Injection,1974-03-30,NaT,560,3245.76,0.0,1817625.60
326153,18537c80-34fc-4171-ab0e-1fdb16ccb0ff,10 ML oxaliplatin 5 MG/ML Injection,1961-07-15,NaT,715,2502.49,0.0,1789280.35
429527,7475a593-3fa7-40c2-bbdf-86c68ebce67b,10 ML oxaliplatin 5 MG/ML Injection,1969-11-22,NaT,549,3003.25,0.0,1648784.25
365081,ae13db8f-fb12-4d5f-bcda-00033532ce87,10 ML oxaliplatin 5 MG/ML Injection,1966-02-27,NaT,660,2480.16,0.0,1636905.60
187004,9c1568ce-b39c-4308-b6ee-6b379a81f623,10 ML oxaliplatin 5 MG/ML Injection,1975-03-10,NaT,550,2893.39,0.0,1591364.50
331128,5113e51b-8403-4613-95f9-5ab2c10c14eb,Acetaminophen 325 MG / oxyCODONE Hydrochloride...,1926-02-03,1990-03-26,780,2027.56,0.0,1581496.80
222528,0336e0b9-6775-4ef2-8a8e-040fdca33d4f,10 ML oxaliplatin 5 MG/ML Injection,1975-01-01,NaT,552,2687.78,0.0,1483654.56
260820,afb52993-2583-4e6f-82cc-28b49fc62051,10 ML oxaliplatin 5 MG/ML Injection,1968-11-26,NaT,627,2300.59,0.0,1442469.93


So the huge TOTALCOST isn't necessarily a data-entry error. It is being driven by a very high number of dispenses combined with a high base cost.

verify the cost relationship

In [20]:
df['calculated_cost'] = df['BASE_COST'] * df['DISPENSES'].head()

In [21]:
df[['BASE_COST', 'DISPENSES', 'TOTALCOST', 'calculated_cost']].head()

,BASE_COST,DISPENSES,TOTALCOST,calculated_cost
0,7.05,1,7.05,7.05
1,5.72,1,5.72,5.72
2,5.25,1,5.25,5.25
3,13.90,1,13.90,13.90
4,263.49,9,2371.41,2371.41


In [22]:
"""Why are we doing this?

We're checking whether:

BASE_COST × DISPENSES ≈ TOTALCOST

For example:

BASE_COST = 2858.12
DISPENSES = 660


2858.12 × 660 ≈ 1,886,359

which matches the huge TOTALCOST."""

"Why are we doing this?\n\nWe're checking whether:\n\nBASE_COST × DISPENSES ≈ TOTALCOST\n\nFor example:\n\nBASE_COST = 2858.12\nDISPENSES = 660\n\n\n2858.12 × 660 ≈ 1,886,359\n\nwhich matches the huge TOTALCOST."

In [23]:
#so as they match remove that calculated column

In [24]:
df=df.drop(columns='calculated_cost')

In [25]:
'calculated_cost' in df

False

In [26]:
df['DESCRIPTION'].nunique()

176

10 most frequently recorded medications

In [27]:
df['DESCRIPTION'].value_counts().head(10)

DESCRIPTION
Hydrochlorothiazide 25 MG Oral Tablet                                                                   53716
insulin human  isophane 70 UNT/ML / Regular Insulin  Human 30 UNT/ML Injectable Suspension [Humulin]    52201
amLODIPine 5 MG / Hydrochlorothiazide 12.5 MG / Olmesartan medoxomil 20 MG Oral Tablet                  36464
Atenolol 50 MG / Chlorthalidone 25 MG Oral Tablet                                                       36170
24 HR Metformin hydrochloride 500 MG Extended Release Oral Tablet                                       32633
Simvastatin 10 MG Oral Tablet                                                                           26313
NDA020503 200 ACTUAT Albuterol 0.09 MG/ACTUAT Metered Dose Inhaler                                      25684
120 ACTUAT Fluticasone propionate 0.044 MG/ACTUAT Metered Dose Inhaler                                  24596
Hydrochlorothiazide 12.5 MG                                                                             1829

In [28]:
medication_utilization = (
    df.groupby('DESCRIPTION')
      .agg(
          total_prescriptions=('DESCRIPTION', 'size'),
          unique_patients=('PATIENT', 'nunique'),
        total_cost=('TOTALCOST', 'sum')
      )
      .sort_values('total_prescriptions', ascending=False)
)

medication_utilization.head(10)

,total_prescriptions,unique_patients,total_cost
DESCRIPTION,,,
Hydrochlorothiazide 25 MG Oral Tablet,53716,1164,1.240395e+08
insulin human isophane 70 UNT/ML / Regular Insulin Human 30 UNT/ML Injectable Suspension [Humulin],52201,388,4.503337e+07
amLODIPine 5 MG / Hydrochlorothiazide 12.5 MG / Olmesartan medoxomil 20 MG Oral Tablet,36464,837,8.467488e+07
Atenolol 50 MG / Chlorthalidone 25 MG Oral Tablet,36170,793,8.210138e+07
24 HR Metformin hydrochloride 500 MG Extended Release Oral Tablet,32633,450,2.648889e+07
Simvastatin 10 MG Oral Tablet,26313,1534,1.553163e+08
NDA020503 200 ACTUAT Albuterol 0.09 MG/ACTUAT Metered Dose Inhaler,25684,1310,9.863031e+07
120 ACTUAT Fluticasone propionate 0.044 MG/ACTUAT Metered Dose Inhaler,24596,596,6.653947e+06
Hydrochlorothiazide 12.5 MG,18291,415,4.152049e+07


In [29]:
"""total_prescriptions → number of medication records
unique_patients → number of different patients receiving it
total_cost → total cost across those records"""

'total_prescriptions → number of medication records\nunique_patients → number of different patients receiving it\ntotal_cost → total cost across those records'

medication cost ranking (Which medications account for the highest total medication spending?)

In [30]:
medication_utilization.sort_values(
    'total_cost',
    ascending=False
).head(10)

,total_prescriptions,unique_patients,total_cost
DESCRIPTION,,,
Simvastatin 10 MG Oral Tablet,26313,1534,1.553163e+08
Hydrochlorothiazide 25 MG Oral Tablet,53716,1164,1.240395e+08
NDA020503 200 ACTUAT Albuterol 0.09 MG/ACTUAT Metered Dose Inhaler,25684,1310,9.863031e+07
amLODIPine 5 MG / Hydrochlorothiazide 12.5 MG / Olmesartan medoxomil 20 MG Oral Tablet,36464,837,8.467488e+07
10 ML oxaliplatin 5 MG/ML Injection,223,223,8.336015e+07
Atenolol 50 MG / Chlorthalidone 25 MG Oral Tablet,36170,793,8.210138e+07
NDA020800 0.3 ML Epinephrine 1 MG/ML Auto-Injector,962,962,6.736424e+07
Vitamin B 12 5 MG/ML Injectable Solution,604,604,5.159582e+07
ferrous sulfate 325 MG Oral Tablet,537,537,4.743321e+07


Cost per prescription

In [31]:
medication_utilization['cost_per_prescription']=medication_utilization['total_cost']/medication_utilization['total_prescriptions']

In [32]:
medication_utilization

,total_prescriptions,unique_patients,total_cost,cost_per_prescription
DESCRIPTION,,,,
Hydrochlorothiazide 25 MG Oral Tablet,53716,1164,1.240395e+08,2309.172285
insulin human isophane 70 UNT/ML / Regular Insulin Human 30 UNT/ML Injectable Suspension [Humulin],52201,388,4.503337e+07,862.691803
amLODIPine 5 MG / Hydrochlorothiazide 12.5 MG / Olmesartan medoxomil 20 MG Oral Tablet,36464,837,8.467488e+07,2322.150146
Atenolol 50 MG / Chlorthalidone 25 MG Oral Tablet,36170,793,8.210138e+07,2269.874926
24 HR Metformin hydrochloride 500 MG Extended Release Oral Tablet,32633,450,2.648889e+07,811.720951
...,...,...,...,...
Penicillin G 375 MG/ML Injectable Solution,2,2,3.735000e+01,18.675000
10 ML Fentanyl 0.05 MG/ML Injection,1,1,2.634900e+02,263.490000
Doxycycline Monohydrate 100 MG Oral Tablet,1,1,3.510900e+02,351.090000


In [33]:
medication_utilization.sort_values(
    'cost_per_prescription',
    ascending=False            #Which medications are expensive per prescription, regardless of how frequently they're prescribed?
).head(10)

,total_prescriptions,unique_patients,total_cost,cost_per_prescription
DESCRIPTION,,,,
10 ML oxaliplatin 5 MG/ML Injection,223,223,83360152.24,373812.341883
Acetaminophen 325 MG / oxyCODONE Hydrochloride 5 MG Oral Tablet,120,117,32210243.31,268418.694250
Tacrine 10 MG Oral Capsule,6,6,906904.21,151150.701667
Acetaminophen/Hydrocodone,54,54,6005991.06,111222.056667
Sertraline 100 MG Oral Tablet,73,73,8108641.26,111077.277534
24hr nicotine transdermal patch,153,153,15018666.51,98161.219020
ferrous sulfate 325 MG Oral Tablet,537,537,47433206.31,88329.993128
Vitamin B 12 5 MG/ML Injectable Solution,604,604,51595821.33,85423.545248
FLUoxetine 20 MG Oral Capsule,71,45,5355697.74,75432.362535


In [34]:
"""Total cost answers → What medications contribute most to overall spending?
Cost per prescription answers → Which medications are expensive each time they are prescribed?"""

'Total cost answers → What medications contribute most to overall spending?\nCost per prescription answers → Which medications are expensive each time they are prescribed?'

Find medications used by the most patients???
Which medications are used by the largest number of unique patients?
This gives us our patient-level medication utilization ranking.


In [35]:
medication_utilization.sort_values('unique_patients',ascending=False).head()

,total_prescriptions,unique_patients,total_cost,cost_per_prescription
DESCRIPTION,,,,
Acetaminophen 500 MG Oral Tablet,1867,1867,9.730686e+05,521.193664
0.4 ML Enoxaparin sodium 100 MG/ML Prefilled Syringe,1790,1790,9.511989e+05,531.396034
Simvastatin 10 MG Oral Tablet,26313,1534,1.553163e+08,5902.644640
NDA020503 200 ACTUAT Albuterol 0.09 MG/ACTUAT Metered Dose Inhaler,25684,1310,9.863031e+07,3840.146141
1 ML Enoxaparin sodium 150 MG/ML Prefilled Syringe,1176,1176,5.559639e+05,472.758418


medication utilization intensity(prescriptions per patient.)

In [36]:
medication_utilization['prescriptions_per_patient']=medication_utilization['total_prescriptions']/medication_utilization['unique_patients']

In [37]:
medication_utilization.sort_values('prescriptions_per_patient',ascending=False).head()  
#High treatment intensity → repeated prescriptions/records per patient(Which medications are prescribed repeatedly to the same patients?)

,total_prescriptions,unique_patients,total_cost,cost_per_prescription,prescriptions_per_patient
DESCRIPTION,,,,,
insulin human isophane 70 UNT/ML / Regular Insulin Human 30 UNT/ML Injectable Suspension [Humulin],52201,388,4.503337e+07,862.691803,134.538660
1 ML Epoetin Alfa 4000 UNT/ML Injection [Epogen],4183,36,1.102179e+06,263.490000,116.194444
Insulin Lispro 100 UNT/ML Injectable Solution [Humalog],307,3,4.305676e+05,1402.500456,102.333333
24 HR Metformin hydrochloride 500 MG Extended Release Oral Tablet,32633,450,2.648889e+07,811.720951,72.517778
Hydrochlorothiazide 25 MG Oral Tablet,53716,1164,1.240395e+08,2309.172285,46.147766


In [38]:
medication_utilization[
    ['total_prescriptions', 'unique_patients', 'prescriptions_per_patient']
].sort_values(
    'unique_patients',
    ascending=False
).head(10)

,total_prescriptions,unique_patients,prescriptions_per_patient
DESCRIPTION,,,
Acetaminophen 500 MG Oral Tablet,1867,1867,1.000000
0.4 ML Enoxaparin sodium 100 MG/ML Prefilled Syringe,1790,1790,1.000000
Simvastatin 10 MG Oral Tablet,26313,1534,17.153194
NDA020503 200 ACTUAT Albuterol 0.09 MG/ACTUAT Metered Dose Inhaler,25684,1310,19.606107
1 ML Enoxaparin sodium 150 MG/ML Prefilled Syringe,1176,1176,1.000000
Hydrochlorothiazide 25 MG Oral Tablet,53716,1164,46.147766
Naproxen sodium 220 MG Oral Tablet,1309,1146,1.142234
NDA020800 0.3 ML Epinephrine 1 MG/ML Auto-Injector,962,962,1.000000
amLODIPine 5 MG / Hydrochlorothiazide 12.5 MG / Olmesartan medoxomil 20 MG Oral Tablet,36464,837,43.565114


In [39]:
#It means:

#Average number of medication records per unique patient.

#A medication record can represent a prescription/claim event, so repeated records don't automatically mean the patient took the medication.

Medication utilization by cost vs frequency(identify medications that are frequently prescribed but also expensive.)

In [40]:
medication_utilization[['total_prescriptions','unique_patients','total_cost','prescriptions_per_patient']].sort_values('total_cost',ascending=False).head(10)
#Which medications contribute most to overall medication spending?

,total_prescriptions,unique_patients,total_cost,prescriptions_per_patient
DESCRIPTION,,,,
Simvastatin 10 MG Oral Tablet,26313,1534,1.553163e+08,17.153194
Hydrochlorothiazide 25 MG Oral Tablet,53716,1164,1.240395e+08,46.147766
NDA020503 200 ACTUAT Albuterol 0.09 MG/ACTUAT Metered Dose Inhaler,25684,1310,9.863031e+07,19.606107
amLODIPine 5 MG / Hydrochlorothiazide 12.5 MG / Olmesartan medoxomil 20 MG Oral Tablet,36464,837,8.467488e+07,43.565114
10 ML oxaliplatin 5 MG/ML Injection,223,223,8.336015e+07,1.000000
Atenolol 50 MG / Chlorthalidone 25 MG Oral Tablet,36170,793,8.210138e+07,45.611602
NDA020800 0.3 ML Epinephrine 1 MG/ML Auto-Injector,962,962,6.736424e+07,1.000000
Vitamin B 12 5 MG/ML Injectable Solution,604,604,5.159582e+07,1.000000
ferrous sulfate 325 MG Oral Tablet,537,537,4.743321e+07,1.000000


Patient-level medication burden
Patient → How many medications/prescriptions does each patient receive?

This is an important healthcare analytics concept because it helps identify high medication burden / potential polypharmacy.

In [41]:
patient_medication_burden=(df.groupby('PATIENT').agg(
                                                   total_prescriptions=('DESCRIPTION', 'size'),
          unique_medications=('DESCRIPTION', 'nunique'),
          total_medication_cost=('TOTALCOST', 'sum')
).sort_values('unique_medications',ascending=False))

In [42]:
patient_medication_burden.head(10)

,total_prescriptions,unique_medications,total_medication_cost
PATIENT,,,
3e5af47c-897c-4ed0-9a58-f94e25eb0366,1095,21,968240.62
b306fd71-d545-47f2-a088-90eb3832024a,294,21,882614.70
02d33b53-1387-4980-8217-2bf6cf05e520,159,19,491968.22
72efe233-6f87-4c9b-ae49-af1e1c8b5fd4,121,19,317377.02
45ab4e6b-37eb-4cce-a3b0-a3128dd3b5f0,76,18,48718.57
c034ceb9-03d8-4b1e-a348-6d1b55e60377,1418,17,791350.33
33c324fb-f552-4c34-b3ec-181e7bf6c328,903,17,579469.11
f59bb525-7cd5-4d99-8b2a-88db5ba6a525,184,17,1116633.71
cbac0349-9a52-4b97-9e7f-bb9072bb10a4,75,17,126035.92


Identify polypharmacy patients..
Now we turn unique_medications into a clinically meaningful category.

For this project, let's use:

< 5 medications → Low medication burden
5–9 → Moderate
≥ 10 → Polypharmacy / high medication burden

In [43]:
import numpy as np

In [44]:
patient_medication_burden['medication_burden_category'] = np.select(
    [
        patient_medication_burden['unique_medications'] < 5,
        patient_medication_burden['unique_medications'].between(5, 9),
        patient_medication_burden['unique_medications'] >= 10
    ],
    [
        'Low',
        'Moderate',
        'Polypharmacy'
    ],default='Unknown'
)

patient_medication_burden['medication_burden_category'].value_counts()

medication_burden_category
Low             6889
Moderate        2349
Polypharmacy     328
Name: count, dtype: int64

Calculate the percentage of patients in each medication-burden category

In [45]:
medication_burden_pct = patient_medication_burden['medication_burden_category'].value_counts(normalize=True).mul(100).round(2)

medication_burden_pct  
#What percentage of patients have polypharmacy-level medication burden?

medication_burden_category
Low             72.02
Moderate        24.56
Polypharmacy     3.43
Name: proportion, dtype: float64

Compare medication burden with medication cost
Do patients with higher medication burden also have higher medication costs?

In [46]:
patient_medication_burden.groupby(
    'medication_burden_category'
).agg(
    patients=('unique_medications', 'size'),
    avg_medications=('unique_medications', 'mean'),
    avg_cost=('total_medication_cost', 'mean'),
    total_cost=('total_medication_cost', 'sum')
).round(2)

,patients,avg_medications,avg_cost,total_cost
medication_burden_category,,,,
Low,6889,2.20,73465.99,5.061072e+08
Moderate,2349,6.28,202523.31,4.757272e+08
Polypharmacy,328,11.69,459695.45,1.507801e+08


In [47]:
patient_medication_burden

,total_prescriptions,unique_medications,total_medication_cost,medication_burden_category
PATIENT,,,,
3e5af47c-897c-4ed0-9a58-f94e25eb0366,1095,21,968240.62,Polypharmacy
b306fd71-d545-47f2-a088-90eb3832024a,294,21,882614.70,Polypharmacy
02d33b53-1387-4980-8217-2bf6cf05e520,159,19,491968.22,Polypharmacy
72efe233-6f87-4c9b-ae49-af1e1c8b5fd4,121,19,317377.02,Polypharmacy
45ab4e6b-37eb-4cce-a3b0-a3128dd3b5f0,76,18,48718.57,Polypharmacy
...,...,...,...,...
bfe3557a-3973-4574-9e68-ca81c1d5c8c4,1,1,2137.38,Low
bfd20809-b9b9-4b6c-9e1f-606d25cddfa3,11,1,31882.29,Low
bfd1f054-8e34-4e5e-b022-748c02911729,1,1,53.65,Low


Does a higher medication burden correspond to higher medication cost?(correlational analysis)

.corr() calculates the correlation between numerical variables.

The result ranges from:

+1 → strong positive relationship
0 → little/no linear relationship
−1 → strong negative relationship

Here we're checking:

Number of unique medications ↔ Total medication cost

In [48]:
patient_medication_burden[['unique_medications','total_medication_cost']].corr()

,unique_medications,total_medication_cost
unique_medications,1.000000,0.516857
total_medication_cost,0.516857,1.000000


In [49]:
""".corr() calculates the Pearson correlation coefficient between the two numerical columns.
How to interpret it

A correlation of +0.52 means there is a moderate positive relationship:

Patients taking a larger number of unique medications tend to have higher total medication costs.

But importantly, correlation does not mean causation.

So don't say:

❌ "More medications cause higher costs."

Say:

✅ "Higher medication burden is moderately associated with higher medication cost."""

'.corr() calculates the Pearson correlation coefficient between the two numerical columns.\nHow to interpret it\n\nA correlation of +0.52 means there is a moderate positive relationship:\n\nPatients taking a larger number of unique medications tend to have higher total medication costs.\n\nBut importantly, correlation does not mean causation.\n\nSo don\'t say:\n\n❌ "More medications cause higher costs."\n\nSay:\n\n✅ "Higher medication burden is moderately associated with higher medication cost.'

High-cost patients(Which patients have the highest total medication spending?)("Which individual patients generate the highest medication costs?")

In [50]:
patient_medication_burden.sort_values('total_medication_cost',ascending=False).head(10)

,total_prescriptions,unique_medications,total_medication_cost,medication_burden_category
PATIENT,,,,
5113e51b-8403-4613-95f9-5ab2c10c14eb,1280,10,2232880.01,Polypharmacy
9d43a914-56cf-463f-8535-a0c0c95fc4a8,37,7,2171508.18,Moderate
0100f99a-1b5d-4a5b-a73f-559a920412e5,13,13,2116117.25,Polypharmacy
18537c80-34fc-4171-ab0e-1fdb16ccb0ff,45,11,2056057.58,Polypharmacy
ae13db8f-fb12-4d5f-bcda-00033532ce87,244,12,2038340.72,Polypharmacy
7475a593-3fa7-40c2-bbdf-86c68ebce67b,153,10,1937443.72,Polypharmacy
c78f5852-5ce3-408b-a355-cf04c0ca635b,39,15,1871026.76,Polypharmacy
0336e0b9-6775-4ef2-8a8e-040fdca33d4f,87,4,1805433.62,Low
3da0bf45-fc25-4005-8ce2-94a178523f2c,157,11,1711016.81,Polypharmacy


In [51]:
patient_medication_burden

,total_prescriptions,unique_medications,total_medication_cost,medication_burden_category
PATIENT,,,,
3e5af47c-897c-4ed0-9a58-f94e25eb0366,1095,21,968240.62,Polypharmacy
b306fd71-d545-47f2-a088-90eb3832024a,294,21,882614.70,Polypharmacy
02d33b53-1387-4980-8217-2bf6cf05e520,159,19,491968.22,Polypharmacy
72efe233-6f87-4c9b-ae49-af1e1c8b5fd4,121,19,317377.02,Polypharmacy
45ab4e6b-37eb-4cce-a3b0-a3128dd3b5f0,76,18,48718.57,Polypharmacy
...,...,...,...,...
bfe3557a-3973-4574-9e68-ca81c1d5c8c4,1,1,2137.38,Low
bfd20809-b9b9-4b6c-9e1f-606d25cddfa3,11,1,31882.29,Low
bfd1f054-8e34-4e5e-b022-748c02911729,1,1,53.65,Low


Do high medication-burden patients also have higher prescription frequency?(compare medication burden vs repeated prescriptions.)

In [52]:
patient_medication_burden[['total_prescriptions','unique_medications']].corr()

,total_prescriptions,unique_medications
total_prescriptions,1.000000,0.441072
unique_medications,0.441072,1.000000


In [53]:
"""So there is a weak-to-moderate positive relationship between:

unique_medications → number of different medications a patient received
total_prescriptions → total medication records/prescriptions for that patient

Patients receiving more unique medications tend to have more prescription records, but the relationship is not particularly strong."""

'So there is a weak-to-moderate positive relationship between:\n\nunique_medications → number of different medications a patient received\ntotal_prescriptions → total medication records/prescriptions for that patient\n\nPatients receiving more unique medications tend to have more prescription records, but the relationship is not particularly strong.'

Which medication-burden category has the highest prescription frequency?

In [54]:
patient_medication_burden.groupby('medication_burden_category').agg(
    patients =('unique_medications','size'),
    avg_prescriptions=('total_prescriptions', 'mean'),
    avg_medications=('unique_medications', 'mean'),
    total_prescriptions=('total_prescriptions', 'sum')
).round(2)

#Polypharmacy patients have substantially higher prescription frequency, averaging about 289 prescription records per patient.

,patients,avg_prescriptions,avg_medications,total_prescriptions
medication_burden_category,,,,
Low,6889,20.67,2.20,142376
Moderate,2349,82.60,6.28,194038
Polypharmacy,328,289.16,11.69,94845


Does medication burden differ in terms of medication spending per patient?
mean = average.

median = middle value after sorting.

Healthcare costs often contain extreme high-cost patients, so median can sometimes represent the typical patient better than mean.

In [55]:
patient_medication_burden.groupby('medication_burden_category').agg(
         patients =('unique_medications','size'),
         mean_cost=('total_medication_cost','mean'),
         median_cost=('total_medication_cost','median'),
         max_cost=('total_medication_cost','max')
).round(2)

#Patients with higher medication burden have substantially higher medication spending.

#The median is especially useful here. For example, the typical Polypharmacy patient has a medication cost of about 375,531, compared with about 32,690 for Low-burden patients.
#Patients with higher medication burden have substantially higher medication spending.

#The median is especially useful here. For example, the typical Polypharmacy patient has a medication cost of about 375,531, compared with about 32,690 for Low-burden patients.

,patients,mean_cost,median_cost,max_cost
medication_burden_category,,,,
Low,6889,73465.99,32689.88,1805433.62
Moderate,2349,202523.31,155888.69,2171508.18
Polypharmacy,328,459695.45,375530.57,2232880.01


Are there patients whose medication cost is unusually high compared with the rest of the population?

In [56]:
patient_medication_burden['total_medication_cost'].describe()

#We already used .describe() on the medication-level data earlier, but now we're applying it to patient-level medication cost.

#This will give us the 25th percentile (Q1), median (Q2), and 75th percentile (Q3), which we'll use to identify high-cost outliers with the IQR method.

count    9.566000e+03
mean     1.184000e+05
std      1.774414e+05
min      4.900000e-01
25%      5.701863e+03
50%      6.180446e+04
75%      1.581157e+05
max      2.232880e+06
Name: total_medication_cost, dtype: float64

Next: calculate the IQR

The IQR (Interquartile Range) measures the spread of the middle 50%:

IQR=Q3−Q1

In [57]:
Q1 = patient_medication_burden['total_medication_cost'].quantile(0.25)
Q3 = patient_medication_burden['total_medication_cost'].quantile(0.75)

IQR = Q3 - Q1

Q1, Q3, IQR

(np.float64(5701.8625), np.float64(158115.7475), np.float64(152413.885))

Next step is to calculate the "upper outlier threshold" and identify patients with unusually high medication costs.
identify the patients above that threshold.


In [58]:
upper_bound = Q3 + (1.5 * IQR)

upper_bound   #This means patients with medication costs above ~386,737 will be considered unusually high-cost patients using the standard IQR outlier rule.

np.float64(386736.575)

In [59]:
high_cost_patients = patient_medication_burden[
    patient_medication_burden['total_medication_cost'] > upper_bound
]

high_cost_patients.sort_values(
    'total_medication_cost',
    ascending=False
).head(10)

#This will show the top 10 unusually high-cost patients, along with their medication burden.

,total_prescriptions,unique_medications,total_medication_cost,medication_burden_category
PATIENT,,,,
5113e51b-8403-4613-95f9-5ab2c10c14eb,1280,10,2232880.01,Polypharmacy
9d43a914-56cf-463f-8535-a0c0c95fc4a8,37,7,2171508.18,Moderate
0100f99a-1b5d-4a5b-a73f-559a920412e5,13,13,2116117.25,Polypharmacy
18537c80-34fc-4171-ab0e-1fdb16ccb0ff,45,11,2056057.58,Polypharmacy
ae13db8f-fb12-4d5f-bcda-00033532ce87,244,12,2038340.72,Polypharmacy
7475a593-3fa7-40c2-bbdf-86c68ebce67b,153,10,1937443.72,Polypharmacy
c78f5852-5ce3-408b-a355-cf04c0ca635b,39,15,1871026.76,Polypharmacy
0336e0b9-6775-4ef2-8a8e-040fdca33d4f,87,4,1805433.62,Low
3da0bf45-fc25-4005-8ce2-94a178523f2c,157,11,1711016.81,Polypharmacy


In [60]:
len(high_cost_patients)

597

Which medication-burden category contributes the most to total medication spending?
see whether the small polypharmacy group accounts for a disproportionately large share of total medication cost.

In [61]:
burden_cost=patient_medication_burden.groupby('medication_burden_category').agg(
    patients =('unique_medications','size'),
    avg_cost=('total_medication_cost','mean'),
    total_cost=('total_medication_cost','sum')
).sort_values('total_cost',ascending=False).round(2)

In [62]:
burden_cost

,patients,avg_cost,total_cost
medication_burden_category,,,
Low,6889,73465.99,5.061072e+08
Moderate,2349,202523.31,4.757272e+08
Polypharmacy,328,459695.45,1.507801e+08


In [63]:
burden_cost['cost_percentage'] = (
    burden_cost['total_cost'] /
    burden_cost['total_cost'].sum()
    * 100
).round(2)

burden_cost

,patients,avg_cost,total_cost,cost_percentage
medication_burden_category,,,,
Low,6889,73465.99,5.061072e+08,44.68
Moderate,2349,202523.31,4.757272e+08,42.00
Polypharmacy,328,459695.45,1.507801e+08,13.31


In [64]:
"""Polypharmacy patients are only 3.43% of patients, but their average medication cost is ~6.3× higher than the Low-burden group.
However, because the polypharmacy group is small, it contributes 13.31% of total spending, not the largest absolute share.
Low-burden patients contribute the largest total spending (44.68%), mainly because there are many more of them.
Moderate-burden patients are also substantial, contributing 42.00%."""

'Polypharmacy patients are only 3.43% of patients, but their average medication cost is ~6.3× higher than the Low-burden group.\nHowever, because the polypharmacy group is small, it contributes 13.31% of total spending, not the largest absolute share.\nLow-burden patients contribute the largest total spending (44.68%), mainly because there are many more of them.\nModerate-burden patients are also substantial, contributing 42.00%.'

Do patients with more medications also have more repeated prescriptions?
This will let us compare medication complexity vs prescription frequency across Low, Moderate, and Polypharmacy patients.

In [65]:
patient_medication_burden.groupby(
    'medication_burden_category'
).agg(
    patients=('unique_medications', 'size'),
    avg_medications=('unique_medications', 'mean'),
    avg_prescriptions=('total_prescriptions', 'mean'),
    median_prescriptions=('total_prescriptions', 'median')
).round(2)
#Patients with higher medication burden also tend to have substantially higher prescription frequency.

,patients,avg_medications,avg_prescriptions,median_prescriptions
medication_burden_category,,,,
Low,6889,2.20,20.67,4.0
Moderate,2349,6.28,82.60,48.0
Polypharmacy,328,11.69,289.16,126.5


High medication burden + unusually high medication cost(Calculate the high-cost threshold and  Identify patients meeting BOTH criteria)
 
 How many patients have both polypharmacy-level medication burden and unusually high medication costs?

In [66]:
high_cost_threshold = Q3 + 1.5 * IQR

high_cost_threshold

np.float64(386736.575)

In [67]:
high_burden_high_cost = patient_medication_burden[
    (patient_medication_burden['unique_medications'] >= 10) &
    (patient_medication_burden['total_medication_cost'] > high_cost_threshold)
].sort_values(
    'total_medication_cost',
    ascending=False
)

high_burden_high_cost.head(10)

,total_prescriptions,unique_medications,total_medication_cost,medication_burden_category
PATIENT,,,,
5113e51b-8403-4613-95f9-5ab2c10c14eb,1280,10,2232880.01,Polypharmacy
0100f99a-1b5d-4a5b-a73f-559a920412e5,13,13,2116117.25,Polypharmacy
18537c80-34fc-4171-ab0e-1fdb16ccb0ff,45,11,2056057.58,Polypharmacy
ae13db8f-fb12-4d5f-bcda-00033532ce87,244,12,2038340.72,Polypharmacy
7475a593-3fa7-40c2-bbdf-86c68ebce67b,153,10,1937443.72,Polypharmacy
c78f5852-5ce3-408b-a355-cf04c0ca635b,39,15,1871026.76,Polypharmacy
3da0bf45-fc25-4005-8ce2-94a178523f2c,157,11,1711016.81,Polypharmacy
ca073c5b-6c05-4ae9-ac65-d567174191ee,883,11,1633468.90,Polypharmacy
fef74b94-c2d9-4e22-bc4f-382c82152126,2845,10,1582859.02,Polypharmacy


In [68]:
len(high_burden_high_cost)

157

In [69]:
high_burden_high_cost_pct = (
    len(high_burden_high_cost) /
    len(patient_medication_burden) *
    100
)

round(high_burden_high_cost_pct, 2)    
  #157 patients have both polypharmacy-level medication burden and unusually high medication cost.
#They represent 1.64% of all patients.

1.64

how much of total medication spending comes from these 157 patients.??

In [70]:
high_burden_high_cost_spending_pct = high_burden_high_cost['total_medication_cost'].sum()/patient_medication_burden['total_medication_cost'].sum()*100
round(high_burden_high_cost_spending_pct,2)

np.float64(10.32)

157 patients met both criteria:
≥10 unique medications (polypharmacy)
Medication cost > 386,736.58 (high-cost outlier threshold)
They represent only 1.64% of all patients.
Yet they account for 10.32% of total medication spending.

A very small group of patients (1.64%) accounts for a disproportionately large share (10.32%) of medication spending.

That is a useful healthcare analytics insight because it identifies a high-cost, high-medication-burden subgroup that could warrant further investigation.

##Which patients have unusually high prescription frequency?

We'll identify patients whose number of prescriptions is an outlier, independent of medication cost.
Do high-burden, high-cost patients have a different medication profile?
identify which medications are most frequently prescribed among those 157 high-burden + high-cost patients.??

In [72]:
# Get the patient IDs of high-burden, high-cost patients
high_burden_patient_ids = high_burden_high_cost.index

# Filter the original medication-level dataframe
high_burden_medications = df[
    df['PATIENT'].isin(high_burden_patient_ids)
]

# Summarize medication utilization
high_burden_medication_profile = (
    high_burden_medications
    .groupby('DESCRIPTION')
    .agg(
        prescriptions=('DESCRIPTION', 'size'),
        unique_patients=('PATIENT', 'nunique'),
        total_cost=('TOTALCOST', 'sum')
    )
    .sort_values('unique_patients', ascending=False)
)

high_burden_medication_profile.head(10)

,prescriptions,unique_patients,total_cost
DESCRIPTION,,,
Simvastatin 10 MG Oral Tablet,2467,86,15023145.84
Warfarin Sodium 5 MG Oral Tablet,5142,82,1248038.00
Verapamil Hydrochloride 40 MG,4654,82,1330085.18
Digoxin 0.125 MG Oral Tablet,5142,82,920837.90
Nitroglycerin 0.4 MG/ACTUAT Mucosal Spray,4172,80,2303961.35
Simvastatin 20 MG Oral Tablet,4000,78,297458.72
Amlodipine 5 MG Oral Tablet,3777,77,429827.16
Clopidogrel 75 MG Oral Tablet,3148,73,550309.96
Acetaminophen 500 MG Oral Tablet,63,63,20815.71


In [73]:
#The key finding is that Simvastatin 10 MG is used by the largest number of high-burden/high-cost patients (86) and also contributes substantially to total spending.

#But notice something important: high prescription frequency ≠ high cost contribution. For example, Warfarin and Digoxin have 5,142 prescriptions each, but their total costs are much lower than Simvastatin 10 MG.

Which medications are disproportionately expensive among high-burden patients?
identify medications where cost is high relative to their prescription frequency.

In [74]:
high_burden_medication_profile['cost_per_prescription']=(high_burden_medication_profile['total_cost']/high_burden_medication_profile['prescriptions']).round(2)
high_burden_medication_profile.sort_values('cost_per_prescription',ascending=False).head(10)

,prescriptions,unique_patients,total_cost,cost_per_prescription
DESCRIPTION,,,,
10 ML oxaliplatin 5 MG/ML Injection,28,28,22070914.45,788246.94
Acetaminophen 325 MG / oxyCODONE Hydrochloride 5 MG Oral Tablet,5,5,3201593.66,640318.73
Tacrine 10 MG Oral Capsule,2,2,727272.62,363636.31
FLUoxetine 20 MG Oral Capsule,1,1,199725.42,199725.42
Carbamazepine[Tegretol],2,2,361736.53,180868.26
Sertraline 100 MG Oral Tablet,13,13,2247833.19,172910.25
Acetaminophen/Hydrocodone,11,11,1796211.33,163291.94
ferrous sulfate 325 MG Oral Tablet,17,17,2525815.14,148577.36
Vitamin B 12 5 MG/ML Injectable Solution,22,22,3142645.23,142847.51


Cost efficiency by medication...combine utilization + cost and identify medications that are both:

frequently prescribed AND expensive (cost_per_patient)


In [75]:
high_burden_medication_profile['cost_per_patient'] = (
    high_burden_medication_profile['total_cost'] /
    high_burden_medication_profile['unique_patients']
).round(2)

high_burden_medication_profile.sort_values(
    ['unique_patients', 'total_cost'],
    ascending=[False, False]
).head(15)

,prescriptions,unique_patients,total_cost,cost_per_prescription,cost_per_patient
DESCRIPTION,,,,,
Simvastatin 10 MG Oral Tablet,2467,86,15023145.84,6089.64,174687.74
Verapamil Hydrochloride 40 MG,4654,82,1330085.18,285.79,16220.55
Warfarin Sodium 5 MG Oral Tablet,5142,82,1248038.00,242.71,15219.98
Digoxin 0.125 MG Oral Tablet,5142,82,920837.90,179.08,11229.73
Nitroglycerin 0.4 MG/ACTUAT Mucosal Spray,4172,80,2303961.35,552.24,28799.52
Simvastatin 20 MG Oral Tablet,4000,78,297458.72,74.36,3813.57
Amlodipine 5 MG Oral Tablet,3777,77,429827.16,113.80,5582.17
Clopidogrel 75 MG Oral Tablet,3148,73,550309.96,174.81,7538.49
Acetaminophen 500 MG Oral Tablet,63,63,20815.71,330.41,330.41


Do patients with both high medication burden and high medication cost also have unusually high prescription frequency?

In [76]:
high_burden_high_cost[['total_prescriptions',
                       'unique_medications',
                       'total_medication_cost']].describe().round(2)

,total_prescriptions,unique_medications,total_medication_cost
count,157.00,157.00,157.00
mean,486.17,12.08,744168.71
std,534.93,2.21,381319.31
min,13.00,10.00,390098.22
25%,140.00,10.00,482967.12
50%,223.00,11.00,618579.58
75%,741.00,13.00,880690.85
max,2845.00,21.00,2232880.01


prescription intensity per medication

This tells us which patients receive repeated prescriptions disproportionately often relative to the number of different medications they use.

In [82]:
high_burden_high_cost['prescriptions_per_medication']=(high_burden_high_cost['total_prescriptions']/high_burden_high_cost['unique_medications']).round(2)

In [87]:
high_burden_high_cost.sort_values('prescriptions_per_medication',ascending=False).head(10)


,total_prescriptions,unique_medications,total_medication_cost,medication_burden_category,prescriptions_per_medication
PATIENT,,,,,
fef74b94-c2d9-4e22-bc4f-382c82152126,2845,10,1582859.02,Polypharmacy,284.50
17a42de1-cd38-42d8-906d-6a21990ced5b,2247,11,710886.93,Polypharmacy,204.27
38967272-1a82-4d94-ac46-9a813f953280,2135,11,1060385.03,Polypharmacy,194.09
edc665ca-8db1-4d11-92ee-3110218766ea,1575,10,697826.96,Polypharmacy,157.50
4d9a197a-0469-4144-9030-0c04654cb4a8,1509,11,856612.61,Polypharmacy,137.18
561d4865-b4ef-4f7b-9fef-e437d2e8a851,1640,12,903018.23,Polypharmacy,136.67
dd5a1f10-b8e6-4141-8e0a-26f34f4ce412,1467,11,882413.23,Polypharmacy,133.36
6f2ff79d-ac7a-4e29-8365-b04912eb3941,1923,15,944882.35,Polypharmacy,128.20
5113e51b-8403-4613-95f9-5ab2c10c14eb,1280,10,2232880.01,Polypharmacy,128.00


High-cost patients with LOW medication burden?
Are there patients with relatively few medications but unusually high medication costs?
This can reveal expensive therapies where cost is driven by the medication itself rather than polypharmacy.

In [89]:
low_burden_high_cost = patient_medication_burden[
    (patient_medication_burden['unique_medications'] <= 5) &
    (patient_medication_burden['total_medication_cost'] > high_cost_threshold)
].sort_values(
    'total_medication_cost',
    ascending=False
)

low_burden_high_cost.head(10)

,total_prescriptions,unique_medications,total_medication_cost,medication_burden_category
PATIENT,,,,
0336e0b9-6775-4ef2-8a8e-040fdca33d4f,87,4,1805433.62,Low
afb52993-2583-4e6f-82cc-28b49fc62051,113,5,1553863.80,Moderate
e70ef6dd-aad9-4f85-8209-3d6ec7ebdecf,3,3,1498903.40,Low
16a9aa74-37da-44ad-83c8-b54eaccdc1d4,31,3,1392661.91,Low
85e986ef-b135-4bd7-84ec-833ddd46fb73,4,4,1370191.59,Low
1391af26-6349-4988-bcdb-9d612d2d161a,16,3,1344274.00,Low
f45956f3-199d-4c68-9322-3f389f0c8c71,4,4,1246033.84,Low
d7fb1226-bc1c-4e0e-9b03-c2e08b05df06,37,4,1218522.07,Low
a095a3c5-a45c-4d48-8a04-8c510d78e60e,4,4,1209411.60,Low


In [93]:
len(low_burden_high_cost)


204

In [94]:
len(patient_medication_burden)

9566

In [95]:
round(
    len(low_burden_high_cost) /
    len(patient_medication_burden) * 100,
    2
)

2.13

Cost concentration / Pareto analysis

Business question:

Are medication costs concentrated among a small proportion of patients?

Step 1 — Sort patients by medication cost

In [98]:
patient_medication_burden.head(2)

,total_prescriptions,unique_medications,total_medication_cost,medication_burden_category
PATIENT,,,,
3e5af47c-897c-4ed0-9a58-f94e25eb0366,1095,21,968240.62,Polypharmacy
b306fd71-d545-47f2-a088-90eb3832024a,294,21,882614.70,Polypharmacy


In [100]:
patient_medication_burden.columns.tolist()

['total_prescriptions',
 'unique_medications',
 'total_medication_cost',
 'medication_burden_category']

In [101]:
cost_concentration = (
    patient_medication_burden
    .reset_index()
    [['PATIENT', 'total_medication_cost']]
    .sort_values('total_medication_cost', ascending=False)
    .reset_index(drop=True)
)

cost_concentration.head(10)

,PATIENT,total_medication_cost
0,5113e51b-8403-4613-95f9-5ab2c10c14eb,2232880.01
1,9d43a914-56cf-463f-8535-a0c0c95fc4a8,2171508.18
2,0100f99a-1b5d-4a5b-a73f-559a920412e5,2116117.25
3,18537c80-34fc-4171-ab0e-1fdb16ccb0ff,2056057.58
4,ae13db8f-fb12-4d5f-bcda-00033532ce87,2038340.72
5,7475a593-3fa7-40c2-bbdf-86c68ebce67b,1937443.72
6,c78f5852-5ce3-408b-a355-cf04c0ca635b,1871026.76
7,0336e0b9-6775-4ef2-8a8e-040fdca33d4f,1805433.62
8,3da0bf45-fc25-4005-8ce2-94a178523f2c,1711016.81
9,ca073c5b-6c05-4ae9-ac65-d567174191ee,1633468.90


Step 2 — Calculate cumulative spending

In [102]:
cost_concentration['cumulative_cost'] = (
    cost_concentration['total_medication_cost'].cumsum()
)

total_cost = cost_concentration['total_medication_cost'].sum()

cost_concentration['cumulative_cost_pct'] = (
    cost_concentration['cumulative_cost'] / total_cost * 100
).round(2)

cost_concentration.head(10)

,PATIENT,total_medication_cost,cumulative_cost,cumulative_cost_pct
0,5113e51b-8403-4613-95f9-5ab2c10c14eb,2232880.01,2232880.01,0.20
1,9d43a914-56cf-463f-8535-a0c0c95fc4a8,2171508.18,4404388.19,0.39
2,0100f99a-1b5d-4a5b-a73f-559a920412e5,2116117.25,6520505.44,0.58
3,18537c80-34fc-4171-ab0e-1fdb16ccb0ff,2056057.58,8576563.02,0.76
4,ae13db8f-fb12-4d5f-bcda-00033532ce87,2038340.72,10614903.74,0.94
5,7475a593-3fa7-40c2-bbdf-86c68ebce67b,1937443.72,12552347.46,1.11
6,c78f5852-5ce3-408b-a355-cf04c0ca635b,1871026.76,14423374.22,1.27
7,0336e0b9-6775-4ef2-8a8e-040fdca33d4f,1805433.62,16228807.84,1.43
8,3da0bf45-fc25-4005-8ce2-94a178523f2c,1711016.81,17939824.65,1.58
9,ca073c5b-6c05-4ae9-ac65-d567174191ee,1633468.90,19573293.55,1.73


Step 3 — Find how much of total spending comes from the top 10% of patients

In [104]:
top_10_pct_count = int(len(cost_concentration) * 0.10)                                           #9,566 patients.

                                                                                             #So:9566×10%=956.6

top_10_cost_pct = (
    cost_concentration.iloc[:top_10_pct_count]['total_medication_cost'].sum()
    / total_cost * 100
)

round(top_10_cost_pct, 2)

np.float64(44.26)

Total patients: 9,566
Top 10% ≈ 956 patients
These 956 highest-cost patients account for 44.26% of total medication spending

So the key finding is:

The top 10% of patients by medication cost account for 44.26% of the total medication spending.

That's a strong indication of cost concentration: a relatively small group of patients contributes a disproportionately large share of medication expenditure.

---What percentage of spending is generated by the top 10% vs top 20% of patients?

In [105]:
top_20_pct_count = int(len(cost_concentration) * 0.20)

top_20_cost_pct = (
    cost_concentration.iloc[:top_20_pct_count]['total_medication_cost'].sum()
    / total_cost * 100
)

round(top_20_cost_pct, 2)

np.float64(64.21)

This is a strong cost-concentration finding.

Just 10% of patients generate 44.26% of total medication spending.
Expanding to the top 20%, they account for 64.21% of spending.
Therefore, the remaining 80% of patients account for only 35.79% of spending.

Medication expenditure was highly concentrated among a relatively small proportion of patients. The top 10% of patients by medication cost accounted for 44.26% of total medication expenditure, while the top 20% accounted for 64.21%.

--Cost vs Prescription Intensity
Do patients with more prescriptions tend to have higher medication costs per prescription?

In [110]:
patient_medication_burden.columns

Index(['total_prescriptions', 'unique_medications', 'total_medication_cost',
       'medication_burden_category', 'cost_per_medication'],
      dtype='str')

In [120]:
patient_medication_burden =patient_medication_burden.drop(columns='cost_per_medication')

In [121]:
patient_medication_burden['cost_per_prescription'] = (
    patient_medication_burden['total_medication_cost']
    / patient_medication_burden['total_prescriptions']
).round(2)

In [122]:
patient_medication_burden.head(5)

,total_prescriptions,unique_medications,total_medication_cost,medication_burden_category,cost_per_prescription
PATIENT,,,,,
3e5af47c-897c-4ed0-9a58-f94e25eb0366,1095,21,968240.62,Polypharmacy,884.24
b306fd71-d545-47f2-a088-90eb3832024a,294,21,882614.70,Polypharmacy,3002.09
02d33b53-1387-4980-8217-2bf6cf05e520,159,19,491968.22,Polypharmacy,3094.14
72efe233-6f87-4c9b-ae49-af1e1c8b5fd4,121,19,317377.02,Polypharmacy,2622.95
45ab4e6b-37eb-4cce-a3b0-a3128dd3b5f0,76,18,48718.57,Polypharmacy,641.03


In [123]:
patient_medication_burden[
    ['total_prescriptions', 'cost_per_prescription']
].corr()

,total_prescriptions,cost_per_prescription
total_prescriptions,1.000000,-0.085418
cost_per_prescription,-0.085418,1.000000


In [124]:
#Patients with more prescriptions do not necessarily have a higher cost per prescription.

#This is actually useful because it tells us that prescription frequency and unit medication cost are largely independent dimensions of medication expenditure.

Which medication-burden category has the greatest cost variability?

How inconsistent is medication spending within each burden category?

In [125]:
burden_cost_variability = (
    patient_medication_burden
    .groupby('medication_burden_category')
    .agg(
        patients=('total_medication_cost', 'size'),
        mean_cost=('total_medication_cost', 'mean'),
        median_cost=('total_medication_cost', 'median'),
        std_cost=('total_medication_cost', 'std'),
        min_cost=('total_medication_cost', 'min'),
        max_cost=('total_medication_cost', 'max')
    )
    .round(2)
)

burden_cost_variability

,patients,mean_cost,median_cost,std_cost,min_cost,max_cost
medication_burden_category,,,,,,
Low,6889,73465.99,32689.88,110660.40,0.49,1805433.62
Moderate,2349,202523.31,155888.69,205424.21,102.97,2171508.18
Polypharmacy,328,459695.45,375530.57,387050.52,2634.90,2232880.01


In [126]:
#Medication expenditure becomes increasingly heterogeneous as medication burden increases, with the polypharmacy group showing the greatest variability in patient-level medication costs.

Which burden category has the highest prescriptions per unique medication?

In [127]:
prescription_intensity = (
    patient_medication_burden
    .groupby('medication_burden_category')
    .agg(
        patients=('total_prescriptions', 'size'),
        avg_prescriptions=('total_prescriptions', 'mean'),
        avg_unique_medications=('unique_medications', 'mean')
    )
)

prescription_intensity['prescriptions_per_medication'] = (
    prescription_intensity['avg_prescriptions'] /
    prescription_intensity['avg_unique_medications']
).round(2)

prescription_intensity.round(2)

,patients,avg_prescriptions,avg_unique_medications,prescriptions_per_medication
medication_burden_category,,,,
Low,6889,20.67,2.20,9.41
Moderate,2349,82.60,6.28,13.16
Polypharmacy,328,289.16,11.69,24.73


So, as medication burden increases, prescriptions are occurring much more frequently relative to the number of distinct medications.

Prescription intensity increased progressively across medication-burden categories, reaching approximately 24.7 prescriptions per unique medication among patients with polypharmacy, compared with 9.4 in the low-burden group.

---How many patients in each medication-burden category have exceptionally high medication costs?

Within each burden category, are there patients whose costs are exceptionally high compared with their peers?

In [128]:
def count_cost_outliers(group):
    q1 = group['total_medication_cost'].quantile(0.25)
    q3 = group['total_medication_cost'].quantile(0.75)
    iqr = q3 - q1
    upper_limit = q3 + 1.5 * iqr
    
    return pd.Series({
        'patients': len(group),
        'upper_cost_limit': round(upper_limit, 2),
        'outlier_patients': (group['total_medication_cost'] > upper_limit).sum()
    })

category_cost_outliers = (
    patient_medication_burden
    .groupby('medication_burden_category')
    .apply(count_cost_outliers)
)

category_cost_outliers

,patients,upper_cost_limit,outlier_patients
medication_burden_category,,,
Low,6889.0,266804.98,320.0
Moderate,2349.0,613508.71,100.0
Polypharmacy,328.0,1203774.43,17.0


In [129]:
"""Conceptually, Pandas does:

Low patients
   ↓
calculate Q1 → Q3 → IQR → upper limit → count outliers

Moderate patients
   ↓
calculate Q1 → Q3 → IQR → upper limit → count outliers

Polypharmacy patients
   ↓
calculate Q1 → Q3 → IQR → upper limit → count outliers"""

'Conceptually, Pandas does:\n\nLow patients\n   ↓\ncalculate Q1 → Q3 → IQR → upper limit → count outliers\n\nModerate patients\n   ↓\ncalculate Q1 → Q3 → IQR → upper limit → count outliers\n\nPolypharmacy patients\n   ↓\ncalculate Q1 → Q3 → IQR → upper limit → count outliers'

Here, each category has its own benchmark.

Low: 320 patients exceed ₹266,804.98.
Moderate: 100 patients exceed ₹613,508.71.
Polypharmacy: 17 patients exceed ₹1,203,774.43.

In [130]:
category_cost_outliers['outlier_pct'] = (
    category_cost_outliers['outlier_patients']
    / category_cost_outliers['patients']
    * 100
).round(2)

category_cost_outliers

,patients,upper_cost_limit,outlier_patients,outlier_pct
medication_burden_category,,,,
Low,6889.0,266804.98,320.0,4.65
Moderate,2349.0,613508.71,100.0,4.26
Polypharmacy,328.0,1203774.43,17.0,5.18


Key finding

Polypharmacy has the highest proportion of unusually high-cost patients: 5.18%.

Approximately 4–5% of patients in each medication-burden category were identified as high-cost outliers using category-specific IQR thresholds, with the highest proportion observed in the polypharmacy group (5.18%).

--->Does having more unique medications correspond to more prescriptions?

In [132]:
medication_diversity_corr = (
    patient_medication_burden[
        ['unique_medications', 'total_prescriptions']
    ]
    .corr()
)

medication_diversity_corr

,unique_medications,total_prescriptions
unique_medications,1.000000,0.441072
total_prescriptions,0.441072,1.000000


In [135]:
"""The correlation between unique medications and total prescriptions is:

r = 0.4411

That indicates a moderate positive correlation.

In simple terms:

Patients who take a greater number of unique medications generally tend to have more prescriptions.

But the relationship is not extremely strong. A correlation of 0.44 means medication count explains only part of the variation in prescription frequency."""

'The correlation between unique medications and total prescriptions is:\n\nr = 0.4411\n\nThat indicates a moderate positive correlation.\n\nIn simple terms:\n\nPatients who take a greater number of unique medications generally tend to have more prescriptions.\n\nBut the relationship is not extremely strong. A correlation of 0.44 means medication count explains only part of the variation in prescription frequency.'

Unique medication count showed a moderate positive correlation with total prescription frequency (r = 0.441), indicating that patients receiving more distinct medications generally had higher prescription volumes, although substantial variation remained unexplained by medication count alone.

-->What proportion of patients account for the highest 50% of medication spending?
finding the minimum patient proportion needed to account for half of total spending.

"What percentage of patients are needed to generate 50% of all medication spending?"

In [149]:
cost_concentration

,total_medication_cost
0,2232880.01
1,2171508.18
2,2116117.25
3,2056057.58
4,2038340.72
...,...
9561,3.43
9562,3.08
9563,2.66
9564,2.20


In [140]:
half_cost_target = total_cost * 0.50
half_cost_target

np.float64(566307269.645)

In [142]:
len(cost_concentration)

9566

In [145]:
patients_for_half_cost = (
    (cost_concentration['cumulative_cost'] >= half_cost_target)   #Finds the first patient where cumulative spending reaches 50%
    .idxmax() + 1)
patients_for_half_cost

1188

In [146]:
half_cost_target = total_cost * 0.50

cost_concentration['cumulative_cost'] = (
    cost_concentration['total_medication_cost'].cumsum()
)

patients_for_half_cost = (
    (cost_concentration['cumulative_cost'] >= half_cost_target)   #Finds the first patient where cumulative spending reaches 50%
    .idxmax() + 1                                                  #.idxmax() finds the first True position.
)

pct_patients_for_half_cost = (
    patients_for_half_cost / len(cost_concentration) * 100
)

round(pct_patients_for_half_cost, 2)                     #Approximately 12.42% of patients account for 50% of total medication spending.

12.42

In simple terms

Think of 100 patients:

~12 patients → 50% of medication spending
~88 patients → remaining 50%

In [148]:
cost_concentration = (
    patient_medication_burden[['total_medication_cost']]
    .sort_values('total_medication_cost', ascending=False)
    .reset_index(drop=True)
)

cost_concentration

,total_medication_cost
0,2232880.01
1,2171508.18
2,2116117.25
3,2056057.58
4,2038340.72
...,...
9561,3.43
9562,3.08
9563,2.66
9564,2.20


Does medication diversity drive spending?
Do patients taking more unique medications tend to have higher total medication costs?

In [152]:
medication_diversity_cost_corr = (
    patient_medication_burden[
        ['unique_medications', 'total_medication_cost']
    ]
    .corr()
)

medication_diversity_cost_corr

,unique_medications,total_medication_cost
unique_medications,1.000000,0.516857
total_medication_cost,0.516857,1.000000


As the number of different medications a patient receives increases, their total medication spending tends to increase as well.

But the relationship is not extremely strong.

A correlation of 0.52 means medication diversity is associated with higher spending, but it is clearly not the only factor determining cost.

"Patients with greater medication diversity tend to have higher total medication costs, with a moderate positive correlation (r = 0.52)."


---examine which burden category contributes the largest share of the patient population and total spending. This will combine patient distribution + spending distribution, rather than another correlation.

In [153]:
burden_category_summary = (
    patient_medication_burden
    .groupby('medication_burden_category')
    .agg(
        patients=('total_medication_cost', 'size'),
        total_cost=('total_medication_cost', 'sum'),
        avg_cost=('total_medication_cost', 'mean')
    )
    .round(2)
)

burden_category_summary

,patients,total_cost,avg_cost
medication_burden_category,,,
Low,6889,5.061072e+08,73465.99
Moderate,2349,4.757272e+08,202523.31
Polypharmacy,328,1.507801e+08,459695.45
